spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkStagedScan.java

```diff
@@ -0,0 +1,114 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.source;
+
+import java.util.List;
+import java.util.Objects;
+import org.apache.iceberg.ScanTask;
+import org.apache.iceberg.ScanTaskGroup;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.spark.ScanTaskSetManager;
+import org.apache.iceberg.spark.SparkReadConf;
+import org.apache.iceberg.util.TableScanUtil;
+import org.apache.spark.sql.SparkSession;
+
+class SparkStagedScan extends SparkScan {
+
+  private final String taskSetID;
+  private final long splitSize;
+  private final int splitLookback;
+  private final long splitOpenFileCost;
+
+  private List<ScanTaskGroup<ScanTask>> taskGroups = null; // lazy cache of tasks
+
+  SparkStagedScan(SparkSession spark, Table table, SparkReadConf readConf) {
+    super(spark, table, readConf, table.schema(), ImmutableList.of());
+
+    this.taskSetID = readConf.scanTaskSetId();
+    this.splitSize = readConf.splitSize();
+    this.splitLookback = readConf.splitLookback();
+    this.splitOpenFileCost = readConf.splitOpenFileCost();
+  }
+
+  protected String taskSetID() {
+    return taskSetID;
+  }
+
+  protected long splitSize() {
+    return splitSize;
+  }
+
+  protected int splitLookback() {
+    return splitLookback;
+  }
+
+  protected long splitOpenFileCost() {
+    return splitOpenFileCost;
+  }
+
+  @Override
+  protected List<ScanTaskGroup<ScanTask>> taskGroups() {
+    if (taskGroups == null) {
+
+      ScanTaskSetManager taskSetManager = ScanTaskSetManager.get();
+      List<ScanTask> tasks = taskSetManager.fetchTasks(table(), taskSetID);
+      ValidationException.check(
+          tasks != null,
+          "Task set manager has no tasks for table %s with id %s",
+          table(),
+          taskSetID);
+
+      this.taskGroups =
+          TableScanUtil.planTaskGroups(tasks, splitSize, splitLookback, splitOpenFileCost);
+    }
+    return taskGroups;
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    }
+
+    if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+
+    SparkStagedScan that = (SparkStagedScan) other;
+    return table().name().equals(that.table().name())
+        && Objects.equals(taskSetID, that.taskSetID())
+        && Objects.equals(splitSize, that.splitSize())
+        && Objects.equals(splitLookback, that.splitLookback())
+        && Objects.equals(splitOpenFileCost, that.splitOpenFileCost());
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(table().name(), taskSetID, splitSize, splitSize, splitOpenFileCost);
+  }
+
+  @Override
+  public String toString() {
+    return String.format(
+        "SparkStagedScan(table=%s, type=%s, taskSetID=%s, caseSensitive=%s)",
```

Comment:
Could you check what naming we use in other scans? Should it be `IcebergStagedScan`?

Reply:
Yea it seems to be the general pattern (though SparkCopyOnWriteScan seems exception).  Changed to IcebergStagedScan